File: leetcode/cpp/13.roman-to-integer.cpp
#include "oj_header.h"

class Solution {
 public:
  /**
   * 字符          数值
   * I             1
   * V             5
   * X             10
   * L             50
   * C             100
   * D             500
   * M             1000
   */
  int romanToInt(string s) { return romanToInt(s, 0); }

  int romanToInt(string& s, int idx) {
    if (s.size() == idx)
      return 0;

    if (s[idx] == 'M')
      return 1000 + romanToInt(s, idx + 1);

    if (s[idx] == 'D')
      return 500 + romanToInt(s, idx + 1);

    if (s[idx] == 'C') {
      if (s.size() > (idx + 1)) {
        if (s[idx + 1] == 'D') {
          return 400 + romanToInt(s, idx + 2);

        } else if (s[idx + 1] == 'M') {
          return 900 + romanToInt(s, idx + 2);
        }
      }
      return 100 + romanToInt(s, idx + 1);
    }

    if (s[idx] == 'L') {
      return 50 + romanToInt(s, idx + 1);
    }

    if (s[idx] == 'X') {
      if (s.size() > (idx + 1)) {
        if (s[idx + 1] == 'L') {
          return 40 + romanToInt(s, idx + 2);

        } else if (s[idx + 1] == 'C') {
          return 90 + romanToInt(s, idx + 2);
        }
      }
      return 10 + romanToInt(s, idx + 1);
    }

    if (s[idx] == 'V') {
      return 5 + romanToInt(s, idx + 1);
    }

    if (s[idx] == 'I') {
      if (s.size() > (idx + 1)) {
        if (s[idx + 1] == 'V') {
          return 4 + romanToInt(s, idx + 2);

        } else if (s[idx + 1] == 'X') {
          return 9 + romanToInt(s, idx + 2);
        }
      }
      return 1 + romanToInt(s, idx + 1);
    }

    return 0;
  }
};

File: leetcode/kt/13.roman-to-integer.kt
class Solution {
    fun romanToInt(s: String): Int {
        /**
         * I             1
         * V             5
         * X             10
         * L             50
         * C             100
         * D             500
         * M             1000
         *
         * I 可以放在 V (5) 和 X (10) 的左边，来表示 4 和 9。
         * X 可以放在 L (50) 和 C (100) 的左边，来表示 40 和 90。
         * C 可以放在 D (500) 和 M (1000) 的左边，来表示 400 和 900。
         */

        if (s.isEmpty())
            return 0;
        return romanToInt(s, 0)
    }

    fun romanToInt(s: String, index: Int): Int {
        val romanChar = s[index]
        var value =
            when (romanChar) {
                'I' -> 1
                'V' -> 5
                'X' -> 10
                'L' -> 50
                'C' -> 100
                'D' -> 500
                'M' -> 1000
                else -> 0
            }
        if(index > 0){
            val prevRomanChar = s[index-1]
            if((romanChar == 'V' || romanChar == 'X') && prevRomanChar == 'I'){
                value -= 2
            }
            if((romanChar == 'L' || romanChar == 'C') && prevRomanChar == 'X'){
                value -= 20
            }
            if((romanChar == 'D' || romanChar == 'M') && prevRomanChar == 'C'){
                value -= 200
            }
        }
        val nextIndex = index + 1
        if (nextIndex >= s.length)
            return value
        return value + romanToInt(s, nextIndex)
    }
}